flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java

```diff
@@ -264,17 +263,32 @@ private void setOwnerReference(Deployment deployment, List<HasMetadata> resource
 	}
 
 	/**
-	 * To get nodePort of configured ports.
+	 * Get rest port from the external Service.
 	 */
-	private int getServiceNodePort(Service service, ConfigOption<Integer> configPort) {
-		final int port = this.flinkConfig.getInteger(configPort);
-		if (service.getSpec() != null && service.getSpec().getPorts() != null) {
-			for (ServicePort p : service.getSpec().getPorts()) {
-				if (p.getPort() == port) {
-					return p.getNodePort();
-				}
-			}
+	private int getRestPortFromExternalService(Service externalService) {
+		final List<ServicePort> servicePortCandidates = externalService.getSpec().getPorts()
+			.stream()
+			.filter(x -> x.getName().equals(Constants.REST_PORT_NAME))
+			.collect(Collectors.toList());
+
+		if (servicePortCandidates.isEmpty()) {
+			throw new RuntimeException("Failed to find port \"" + Constants.REST_PORT_NAME + "\" in Service \"" +
+				KubernetesUtils.getRestServiceName(this.clusterId) + "\"");
+		}
+
+		final ServicePort externalServicePort = servicePortCandidates.get(0);
+
+		final KubernetesConfigOptions.ServiceExposedType externalServiceType =
+			KubernetesConfigOptions.ServiceExposedType.valueOf(externalService.getSpec().getType());
+
+		switch (externalServiceType) {
+			case ClusterIP:
+			case LoadBalancer:
```

Comment:
I think we change the behavior here. When the `REST_SERVICE_EXPOSED_TYPE` is `LoadBalancer` and the cluster do not provide the loadbalancer, then a NodePort url should be returned. After this change, it could not work now.

This change will also make the K8s session e2e tests failed.

@zhengcanbin @TisonKun Could you please have a look?

Reply:
The azure pipeline passed just because we have disabled the K8s related e2e tests. If you run in an unmanaged K8s cluster without load balancer support, then it could not work now.

Reply:
If the cluster does not provide the LoadBalancer, I think we should throw Exception instead of return the NodePort URL. Can we separate the Service type more clearly?

Reply:
Since `LoadBalancer` is a superset of `NodePort`, so when the service exposed type is `LoadBalancer`, K8s will always create `NodePort`. In my opinion, returning a node port url when the `LoadBalancer` is not available is a reasonable solution. And i believe that many users are running in unmanaged K8s cluster without load balancer. Now they have to explicitly set the `kubernetes.rest-service.exposed.type=NodePort`.

Reply:
I have no idea if you really want to completely separate the `LoadBalancer` and `NodePort`. However, in such case, we should wait for the load balancer ready and return the `EXTERNAL-IP:8081` as JobManager url. Otherwise, the Flink client will timeout and should clean-up the cluster resources.

Then for unmanaged K8s cluster without load balancer, we will enforce the users to set `kubernetes.rest-service.exposed.type=NodePort` explicitly, rather that return a confusing JobManager url `MASTER_ADDRESS:8081`. Also the submission will always fail with timeout.

Reply:
> Since `LoadBalancer` is a superset of `NodePort`, so when the service exposed type is `LoadBalancer`, K8s will always create `NodePort`.

Yes, you are right!

> In my opinion, returning a node port url when the `LoadBalancer` is not available is a reasonable solution. And i believe that many users are running in unmanaged K8s cluster without load balancer. Now they have to explicitly set the `kubernetes.rest-service.exposed.type=NodePort`.

I tend to think differently. If people use the LB Service, then the LB should work as expected, otherwise, we should throw Exceptions when retrieving the Endpoint to tell the users that the LB is unready or abnormal. 
Imagine what could happen if we fall back to the NodePort in such a situation, the jobs may never be submitted to the cluster since the Kubernetes node IP/address are not accessible in the submission client due to network security policy, which could be the usual cases, especially on Cloud. 
Therefore, such toleration or fallback could confuse the users and they may need to dive into the code to get what really happened behind.
If the users run their workload on K8s clusters without LB, then it is reasonable for them to explicitly set the `kubernetes.rest-service.exposed.type=NodePort`.

Another concern, maybe we'd better change the default value of the  `kubernetes.rest-service.exposed.type` to NodePort or ClusterIP.

Reply:
I know your concern and i do not insist on the fallback logics. What i am talking is how to make separating `LoadBalancer` and `NodePort` more reasonable. 

If the user set the `kubernetes.rest-service.exposed.type=LoadBalancer`, then we should wait for the load balancer ready or timeout, just as you say, we should not print a confusing JobManager url in the log. This is the current problem.

About the default value of `kubernetes.rest-service.exposed.type`, it needs a wider discussion.

Reply:
> However, in such case, we should wait for the load balancer ready and return the `EXTERNAL-IP:8081` as JobManager url. Otherwise, the Flink client will timeout and should clean-up the cluster resources.

Currently, we clean up the K8s resources once errors happen during the creation of the `Deployment`, `ConfigMap`, and `Service`; we throw an Exception instead of cleaning up all the resources if we failed to retrieve the Endpoint, which I think is reasonable since the external Service is a bypass to some extent.

Regarding whether the `Service` is ready, previously I left a question at https://github.com/apache/flink/pull/11233, I will file another ticket for further discussion and fixup.

> Then for unmanaged K8s cluster without load balancer, we will enforce the users to set kubernetes.rest-service.exposed.type=NodePort explicitly, rather that return a confusing JobManager url MASTER_ADDRESS:8081. Also the submission will always fail with timeout.

Sorry that I don't understand the core problem here. It would be nice if you provide more information.

Lastly, I was planning to rework the implementation of `Fabric8FlinkKubeClient#getRestEndpoint` to clearly separating the functionality of `NodePort` and `LB` Service, meanwhile fix some bugs in that method.

Reply:
Glad to hear that you will create a new ticket to rework the `getRestEndpoint` according to `LoadBalacer` or `NodePort`. I think this change makes sense to me.

Currently, the problem in the unmanaged K8s cluster without LD is Flink client print a wrong JobManager url, which is in such schema `K8s_Master:8081`, it should be `K8s_Master:NodePort`. I think this could be fixed in your upcoming tickets.

Reply:
> Glad to hear that you will create a new ticket to rework the `getRestEndpoint` according to `LoadBalacer` or `NodePort`. I think this change makes sense to me.
> 
> Currently, the problem in the unmanaged K8s cluster without LD is Flink client print a wrong JobManager url, which is in such schema `K8s_Master:8081`, it should be `K8s_Master:NodePort`. I think this could be fixed in your upcoming tickets.

The fix for this problem is in the plan.

Reply:
Has this been resolved yet